### PR TITLE
homework Shapes

### PR DESCRIPTION
**Pull request by KGebski0036**

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
:+1:

**Review by WBigaj — APPROVED** [review]
I have checked code and it looks good :)

File: shapes/Square.cpp [hunk 1]
```diff
@@ -1,11 +1,9 @@
 #include "Square.hpp"
 #include <iostream>
 
-Square::Square(Color c) : Rectangle(c) {}
+Square::Square(double x, Color color)
+    : Rectangle(x, x, color) {}
```
Comment (ziobron) [inline shapes/Square.cpp:5]: Please also add a delegating constructor in Square

File: shapes/Rectangle.cpp [hunk 2]
```diff
@@ -1,9 +1,10 @@
 #include "Rectangle.hpp"
 #include <iostream>
 
-Rectangle::Rectangle(double x, double y)
+Rectangle::Rectangle(double x, double y, Color color)
     : x_(x),
-      y_(y)
+      y_(y),
+      Shape(color)
 {}
```
Comment (ziobron) [inline shapes/Rectangle.cpp:8]: Please also add a delegating constructor in Rectangle

File: shapes/main.cpp [hunk 3]
```diff
@@ -1,96 +1,95 @@
...
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
-    return false;
+std::function<bool(shared_ptr<Shape>)> areaLessThanX{[x{10}](shared_ptr<Shape> s) {
```
Comment (ziobron) [inline shapes/main.cpp:33]: auto here as a type is enough. You only need to pass it to findFirstShapeMatchingPredicate as std::function.

File: shapes/main.cpp [hunk 4]
```diff
@@ -1,96 +1,95 @@
...
+}};
+
+template <typename T>
+using IsDerivedFromShape = std::enable_if_t<std::is_base_of_v<Shape, std::remove_reference_t<T>>>;
+
+template <typename T,
+          typename = IsDerivedFromShape<T>>
+void printShape(T&& shape) {
+    shape.print();
 }
```
Comment (ziobron) [inline shapes/main.cpp:44]: this does not do what you was asked for:
> Write a function that allows inserting only subclasses of Shape to the collection

**Review by ziobron — CHANGES_REQUESTED** [review]

File: shapes/Shape.hpp [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
     virtual ~Shape() {}
 
-    virtual auto getArea() const -> double = 0;
+    virtual double getArea() const = 0;
     virtual double getPerimeter() const = 0;
     virtual void print() const;
+    virtual Color getColor() const { return color_; }
+
+private:
+    Color color_ = Color::Red;
```
Comment (niemarcin) [inline shapes/Shape.hpp:23]:
```suggestion
    Color color_{Color::Red};
```
Should work fine as well :)

**Review by niemarcin — COMMENTED** [review]
Minor suggestion added

**Comment by ziobron** [comment]
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4448, 4450, 4451, 4452] }

**Pull request closed without merging**